reframe/frontend/executors/__init__.py

```diff
@@ -162,15 +209,23 @@ def _notify_listeners(self, callback_name):
     def _safe_call(self, fn, *args, **kwargs):
         if fn.__name__ != 'poll':
             self._current_stage = fn.__name__
+            self._timestamps[f'{self._current_stage}_start'] = time.time()
 
         try:
             with logging.logging_context(self.check) as logger:
                 logger.debug('entering stage: %s' % self._current_stage)
-                return fn(*args, **kwargs)
+                ret = fn(*args, **kwargs)
+                self._timestamps[f'{self._current_stage}_finish'] = time.time()
+                self._timestamps['pipeline_end'] = time.time()
+                return ret
```

Comment:
We can avoid the duplication for each case, by using a context manager to do this thing. It doesn't have to be public, just define it nested in this function. Then you can simply write:

```python
with update_timestamps():
    return fn(*args, **kwargs)
```